In [38]:
import numpy as np
import matplotlib.pyplot as plt

In [39]:
def calc_mae(y, y_pred):
    err = np.mean(np.abs(y - y_pred))
    return err

def calc_mse(y, y_pred):
    err = np.mean((y - y_pred)**2)
    return err

###  Задание 1
Подберите скорость обучения (alpha) и количество итераций

In [40]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]
n = X.shape[0]

eta = 0.1 
n_iter = 1000

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 10        
Learning rate = 0.1        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [11.8 38.2], MSE = 3047.75
Iteration #10: W_new = [12651.73553914 69617.0969639 ], MSE = 18310954068.05
Iteration #20: W_new = [ 7732434.81888021 42641607.37852182], MSE = 9128819654907552.0
Iteration #30: W_new = [1.06344502e+09 5.86454589e+09], MSE = 2.327920364266843e+20
Iteration #40: W_new = [3.00127077e+10 1.65510116e+11], MSE = 2.545133529815938e+23
Iteration #50: W_new = [1.55345341e+11 8.56677968e+11], MSE = 9.572295620500076e+24
Iteration #60: W_new = [1.27742291e+11 7.04456313e+11], MSE = 9.351480126475979e+24
Iteration #70: W_new = [1.38141953e+10 7.61806995e+10], MSE = 1.6408589528283872e+23
Iteration #80: W_new = [1.51674189e+08 8.36432543e+08], MSE = 3.125533537874452e+19
Iteration #90: W_new = [116395.49984147 641638.79865015], MSE = 31317286806394.18
Iteration #100: W_new = [48.41963663 22.99884414], MSE = 53518.86
Iteration #110: W_new = [44.97714033  3.82798

In [41]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]
n = X.shape[0]

eta = 0.05 
n_iter = 1000

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 10        
Learning rate = 0.05        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [ 6.4  19.35], MSE = 3047.75
Iteration #10: W_new = [18.77319138  8.6270072 ], MSE = 398.58
Iteration #20: W_new = [27.15289315  7.06014716], MSE = 207.1
Iteration #30: W_new = [32.4241672   6.10426679], MSE = 124.63
Iteration #40: W_new = [35.85213314  5.48265802], MSE = 86.53
Iteration #50: W_new = [38.15155609  5.06569312], MSE = 67.79
Iteration #60: W_new = [39.73790007  4.7780341 ], MSE = 58.04
Iteration #70: W_new = [40.86051023  4.57446606], MSE = 52.69
Iteration #80: W_new = [41.67349173  4.4270444 ], MSE = 49.62
Iteration #90: W_new = [42.2746909  4.3180262], MSE = 47.78
Iteration #100: W_new = [42.727794    4.23586294], MSE = 46.63
Iteration #110: W_new = [43.07521157  4.17286412], MSE = 45.89
Iteration #120: W_new = [43.34578807  4.12379924], MSE = 45.4
Iteration #130: W_new = [43.55952748  4.08504089], MSE = 45.06
Iteration #140: W_new = [43.73055537  4.05402762],

In [42]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]
n = X.shape[0]

eta = 0.01 
n_iter = 1000

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 10        
Learning rate = 0.01        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [2.08 4.27], MSE = 3047.75
Iteration #10: W_new = [ 6.67106886 10.61676385], MSE = 749.71
Iteration #20: W_new = [ 9.49320908 10.25731657], MSE = 648.91
Iteration #30: W_new = [11.85740092  9.83349244], MSE = 570.46
Iteration #40: W_new = [13.86876921  9.46898661], MSE = 508.03
Iteration #50: W_new = [15.59085668  9.15672679], MSE = 457.73
Iteration #60: W_new = [17.07337653  8.88789585], MSE = 416.77
Iteration #70: W_new = [18.35601294  8.65530964], MSE = 383.06
Iteration #80: W_new = [19.47073522  8.45317196], MSE = 355.08
Iteration #90: W_new = [20.44350656  8.27677488], MSE = 331.65
Iteration #100: W_new = [21.29557245  8.12226587], MSE = 311.9
Iteration #110: W_new = [22.044442    7.98646992], MSE = 295.12
Iteration #120: W_new = [22.7046421   7.86675281], MSE = 280.78
Iteration #130: W_new = [23.2883015   7.76091518], MSE = 268.46
Iteration #140: W_new = [23.80560705  7

In [43]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]
n = X.shape[0]

eta = 0.005 
n_iter = 1000

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 10        
Learning rate = 0.005        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [1.54  2.385], MSE = 3047.75
Iteration #10: W_new = [4.59259286 9.51117298], MSE = 875.43
Iteration #20: W_new = [ 6.30924967 10.4974107 ], MSE = 757.19
Iteration #30: W_new = [ 7.66009887 10.51154395], MSE = 707.03
Iteration #40: W_new = [ 8.8189058  10.36147111], MSE = 666.4
Iteration #50: W_new = [ 9.83480137 10.19316428], MSE = 631.75
Iteration #60: W_new = [10.73192539 10.03523492], MSE = 601.93
Iteration #70: W_new = [11.52711486  9.89261853], MSE = 576.14
Iteration #80: W_new = [12.23380578  9.76504901], MSE = 553.73
Iteration #90: W_new = [12.86321242  9.65114656], MSE = 534.18
Iteration #100: W_new = [13.42485553  9.54940084], MSE = 517.06
Iteration #110: W_new = [13.92688737  9.45841124], MSE = 502.02
Iteration #120: W_new = [14.37632807  9.37693487], MSE = 488.76
Iteration #130: W_new = [14.77925036  9.30388303], MSE = 477.05
Iteration #140: W_new = [15.14092825  

In [44]:
X = np.array([[ 1,  1],
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
y = [45, 55, 50, 55, 60, 35, 75, 80, 50, 60]
n = X.shape[0]

eta = 0.001 
n_iter = 1000

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
    for k in range(W.shape[0]):
        W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    if i % 10 == 0:
        eta /= 1.1
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err, 2)}')

Number of objects = 10        
Learning rate = 0.001        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [1.108 0.877], MSE = 3047.75
Iteration #10: W_new = [1.98538229 3.73754631], MSE = 1991.2
Iteration #20: W_new = [2.65151474 5.62717924], MSE = 1473.64
Iteration #30: W_new = [3.17602393 6.90966203], MSE = 1203.1
Iteration #40: W_new = [3.60140403 7.80143614], MSE = 1052.67
Iteration #50: W_new = [3.95459332 8.43516509], MSE = 963.95
Iteration #60: W_new = [4.25330617 8.89440055], MSE = 908.59
Iteration #70: W_new = [4.50960937 9.23308795], MSE = 872.15
Iteration #80: W_new = [4.73200372 9.48685965], MSE = 846.97
Iteration #90: W_new = [4.9266705  9.67974823], MSE = 828.76
Iteration #100: W_new = [5.09823829 9.82827603], MSE = 815.07
Iteration #110: W_new = [5.25026691 9.94400351], MSE = 804.43
Iteration #120: W_new = [ 5.38556058 10.03515101], MSE = 795.91
Iteration #130: W_new = [ 5.50637544 10.10765119], MSE = 788.93
Iteration #140: W_new = [ 5.61456026 10.16584453], MSE 

Вывод: при большей скорости обучения, требуется меньшее количество итераций. При скорости обучения 0.1 количество итераций составляет примерно 110 шагов. При  скорости обучения 0.05 количество итераций составляет примерно 700 шагов.

### Задание 2 
В этом коде мы избавляемся от итераций по весам, но тут есть ошибка, исправьте ее

In [45]:
n = X.shape[0]

eta = 1e-2 
n_iter = 100

W = np.array([1, 0.5])
print(f'Number of objects = {n} \
       \nLearning rate = {eta} \
       \nInitial weights = {W} \n')

for i in range(n_iter):
    y_pred = np.dot(X, W)
    err = calc_mse(y, y_pred)
#     for k in range(W.shape[0]):
#         W[k] -= eta * (1/n * 2 * X[:, k] @ (y_pred - y))
    # ИЗМЕНЕНИЯ: указываем столбец в матрице Х
    W -= eta * (1/n * 2 * np.dot(X[:, 1], y_pred - y))
    # ИЗМЕНЕНИЯ - добавлена строка с изменением шага(eta)
    eta /= 1.1
    if i % 10 == 0:
        print(f'Iteration #{i}: W_new = {W}, MSE = {round(err,2)}')

Number of objects = 10        
Learning rate = 0.01        
Initial weights = [1.  0.5] 

Iteration #0: W_new = [4.77 4.27], MSE = 3047.75
Iteration #10: W_new = [10.10145365  9.60145365], MSE = 624.54
Iteration #20: W_new = [10.30592861  9.80592861], MSE = 613.07
Iteration #30: W_new = [10.3429761  9.8429761], MSE = 611.47
Iteration #40: W_new = [10.35382943  9.85382943], MSE = 611.02
Iteration #50: W_new = [10.35760203  9.85760203], MSE = 610.87
Iteration #60: W_new = [10.35900004  9.85900004], MSE = 610.82
Iteration #70: W_new = [10.35953089  9.85953089], MSE = 610.8
Iteration #80: W_new = [10.35973435  9.85973435], MSE = 610.79
Iteration #90: W_new = [10.35981262  9.85981262], MSE = 610.79


### Задание 3 
Вместо того чтобы задавать количество итераций, задайте условие остановки алгоритма, когда ошибка за итерацию начинает изменяться ниже определённого порога — упрощённый аналог параметра tol в линейной регрессии в sklearn.

In [46]:
def mserror(X, w, y):
    y_pred = X.dot(w)
    return (np.sum((y_pred - y)**2)) / len(y)
X = np.array([[ 1,  1],              
              [ 1,  1],
              [ 1,  2],
              [ 1,  5],
              [ 1,  3],
              [ 1,  0],
              [ 1,  5],
              [ 1, 10],
              [ 1,  1],
              [ 1,  2]])
Y = np.array([45, 55, 50, 55, 60, 35, 75, 80, 50, 60])
W = np.array([1, 0.5])
weight_dist = np.inf
min_weight_dist = 1e-8
eta = 0.01
errors = []
iter_num = 0

while weight_dist > min_weight_dist:
    y_pred = np.dot(X, W)
    dQ = 2 / Y.shape[0] * np.dot(X.T, y_pred - Y)
    new_w = W - eta * dQ
    weight_dist = np.linalg.norm(new_w - W, ord=2)
    error = mserror(X, new_w, Y)        
    errors.append(error)
    
    print(f'Iter {iter_num}: error - {error}, weights: {new_w}')
    
    iter_num += 1
    W = new_w   

print(f'В случае использования градиентного спуска функционал ошибки составляет {round(errors[-1], 4)}')

Iter 0: error - 1777.0353, weights: [2.08 4.27]
Iter 1: error - 1233.1246862399998, weights: [2.9122 6.6934]
Iter 2: error - 995.6119136616959, weights: [3.582352 8.242912]
Iter 3: error - 887.3532060945618, weights: [4.14613024 9.2253808 ]
Iter 4: error - 833.705103731704, weights: [4.63968479 9.83998351]
Iter 5: error - 803.2133893510008, weights: [ 5.08649208 10.21600803]
Iter 6: error - 782.6219704436712, weights: [ 5.50180176 10.43737578]
Iter 7: error - 766.3427039405008, weights: [ 5.89552318 10.55855991]
Iter 8: error - 752.0184356756685, weights: [ 6.27409912 10.61491815]
Iter 9: error - 738.6534565814403, weights: [ 6.64172205 10.62940003]
Iter 10: error - 725.8257570810141, weights: [ 7.0011236 10.6169007]
Iter 11: error - 713.3551043907684, weights: [ 7.35408709 10.58708704]
Iter 12: error - 701.1631838639862, weights: [ 7.70178013 10.54623222]
Iter 13: error - 689.2146678156332, weights: [ 8.04497059 10.49840646]
Iter 14: error - 677.4923836960982, weights: [ 8.38416679 10

Iter 240: error - 54.0448971282921, weights: [40.43718613  4.65122936]
Iter 241: error - 53.861947777931675, weights: [40.47936864  4.64358021]
Iter 242: error - 53.68232017986122, weights: [40.52116646  4.63600082]
Iter 243: error - 53.50595402209676, weights: [40.56258308  4.62849055]
Iter 244: error - 53.33279008771971, weights: [40.60362198  4.62104878]
Iter 245: error - 53.16277023499373, weights: [40.64428662  4.61367488]
Iter 246: error - 52.995837377843415, weights: [40.68458039  4.60636822]
Iter 247: error - 52.83193546668673, weights: [40.72450669  4.5991282 ]
Iter 248: error - 52.671009469615726, weights: [40.76406887  4.59195421]
Iter 249: error - 52.5130053539192, weights: [40.80327023  4.58484565]
Iter 250: error - 52.35787006794029, weights: [40.84211409  4.57780191]
Iter 251: error - 52.205551523264056, weights: [40.88060369  4.57082242]
Iter 252: error - 52.055998577228, weights: [40.91874228  4.56390657]
Iter 253: error - 51.909161015750485, weights: [40.95653304  4.5

Iter 515: error - 44.03404757054602, weights: [44.69015762  3.88001855]
Iter 516: error - 44.0328619836387, weights: [44.69355335  3.87940279]
Iter 517: error - 44.0316979230501, weights: [44.69691812  3.87879264]
Iter 518: error - 44.030554997933805, weights: [44.7002522   3.87818805]
Iter 519: error - 44.029432824539825, weights: [44.70355587  3.87758898]
Iter 520: error - 44.028331026085894, weights: [44.70682942  3.87699538]
Iter 521: error - 44.027249232630766, weights: [44.7100731   3.87640718]
Iter 522: error - 44.02618708095016, weights: [44.71328721  3.87582435]
Iter 523: error - 44.0251442144147, weights: [44.71647201  3.87524684]
Iter 524: error - 44.02412028287025, weights: [44.71962776  3.8746746 ]
Iter 525: error - 44.02311494252028, weights: [44.72275472  3.87410757]
Iter 526: error - 44.02212785581047, weights: [44.72585318  3.87354571]
Iter 527: error - 44.02115869131536, weights: [44.72892337  3.87298898]
Iter 528: error - 44.02020712362711, weights: [44.73196556  3.8

Iter 781: error - 43.96924902215495, weights: [45.02994978  3.81840248]
Iter 782: error - 43.9692399615708, weights: [45.03024664  3.81834865]
Iter 783: error - 43.96923106549674, weights: [45.03054079  3.81829531]
Iter 784: error - 43.96922233094581, weights: [45.03083225  3.81824246]
Iter 785: error - 43.96921375498533, weights: [45.03112106  3.81819009]
Iter 786: error - 43.96920533473579, weights: [45.03140723  3.81813819]
Iter 787: error - 43.96919706736998, weights: [45.0316908   3.81808677]
Iter 788: error - 43.96918895011209, weights: [45.03197178  3.81803582]
Iter 789: error - 43.96918098023661, weights: [45.03225019  3.81798534]
Iter 790: error - 43.9691731550676, weights: [45.03252607  3.81793531]
Iter 791: error - 43.96916547197764, weights: [45.03279943  3.81788574]
Iter 792: error - 43.96915792838709, weights: [45.03307029  3.81783662]
Iter 793: error - 43.96915052176307, weights: [45.03333869  3.81778795]
Iter 794: error - 43.96914324961872, weights: [45.03360464  3.8177

Iter 1068: error - 43.96875259554548, weights: [45.06015249  3.81292569]
Iter 1069: error - 43.96875254841899, weights: [45.06017389  3.8129218 ]
Iter 1070: error - 43.96875250214817, weights: [45.06019511  3.81291796]
Iter 1071: error - 43.968752456717475, weights: [45.06021613  3.81291414]
Iter 1072: error - 43.968752412111655, weights: [45.06023696  3.81291037]
Iter 1073: error - 43.96875236831572, weights: [45.0602576   3.81290663]
Iter 1074: error - 43.968752325314966, weights: [45.06027805  3.81290292]
Iter 1075: error - 43.968752283094986, weights: [45.06029831  3.81289924]
Iter 1076: error - 43.96875224164156, weights: [45.06031839  3.8128956 ]
Iter 1077: error - 43.96875220094079, weights: [45.06033829  3.81289199]
Iter 1078: error - 43.96875216097903, weights: [45.060358    3.81288842]
Iter 1079: error - 43.96875212174283, weights: [45.06037754  3.81288488]
Iter 1080: error - 43.96875208321903, weights: [45.06039689  3.81288137]
Iter 1081: error - 43.968752045394695, weights:

Iter 1270: error - 43.96875006408423, weights: [45.06213113  3.81256689]
Iter 1271: error - 43.968750062920684, weights: [45.0621345   3.81256628]
Iter 1272: error - 43.968750061778245, weights: [45.06213783  3.81256567]
Iter 1273: error - 43.968750060656575, weights: [45.06214113  3.81256507]
Iter 1274: error - 43.96875005955524, weights: [45.06214441  3.81256448]
Iter 1275: error - 43.9687500584739, weights: [45.06214765  3.81256389]
Iter 1276: error - 43.968750057412215, weights: [45.06215086  3.81256331]
Iter 1277: error - 43.968750056369814, weights: [45.06215405  3.81256273]
Iter 1278: error - 43.96875005534632, weights: [45.0621572   3.81256216]
Iter 1279: error - 43.96875005434143, weights: [45.06216033  3.81256159]
Iter 1280: error - 43.96875005335475, weights: [45.06216343  3.81256103]
Iter 1281: error - 43.96875005238601, weights: [45.0621665   3.81256048]
Iter 1282: error - 43.96875005143484, weights: [45.06216954  3.81255992]
Iter 1283: error - 43.96875005050098, weights: 

Iter 1407: error - 43.96875000520633, weights: [45.06239486  3.81251907]
Iter 1408: error - 43.968750005111794, weights: [45.06239582  3.81251889]
Iter 1409: error - 43.96875000501898, weights: [45.06239677  3.81251872]
Iter 1410: error - 43.96875000492787, weights: [45.06239771  3.81251855]
Iter 1411: error - 43.96875000483838, weights: [45.06239865  3.81251838]
Iter 1412: error - 43.96875000475053, weights: [45.06239957  3.81251821]
Iter 1413: error - 43.96875000466428, weights: [45.06240049  3.81251805]
Iter 1414: error - 43.9687500045796, weights: [45.06240139  3.81251788]
Iter 1415: error - 43.96875000449644, weights: [45.06240229  3.81251772]
Iter 1416: error - 43.968750004414794, weights: [45.06240318  3.81251756]
Iter 1417: error - 43.968750004334645, weights: [45.06240407  3.8125174 ]
Iter 1418: error - 43.96875000425594, weights: [45.06240494  3.81251724]
Iter 1419: error - 43.96875000417866, weights: [45.06240581  3.81251708]
Iter 1420: error - 43.9687500041028, weights: [45

Iter 1623: error - 43.968750000099455, weights: [45.06248547  3.81250264]
Iter 1624: error - 43.96875000009765, weights: [45.0624856   3.81250261]
Iter 1625: error - 43.96875000009588, weights: [45.06248573  3.81250259]
Iter 1626: error - 43.96875000009415, weights: [45.06248586  3.81250256]
Iter 1627: error - 43.968750000092435, weights: [45.06248599  3.81250254]
Iter 1628: error - 43.968750000090736, weights: [45.06248612  3.81250252]
Iter 1629: error - 43.9687500000891, weights: [45.06248625  3.81250249]
Iter 1630: error - 43.96875000008749, weights: [45.06248637  3.81250247]
Iter 1631: error - 43.968750000085905, weights: [45.0624865   3.81250245]
Iter 1632: error - 43.96875000008434, weights: [45.06248662  3.81250243]
Iter 1633: error - 43.968750000082814, weights: [45.06248674  3.8125024 ]
Iter 1634: error - 43.9687500000813, weights: [45.06248686  3.81250238]
Iter 1635: error - 43.96875000007983, weights: [45.06248698  3.81250236]
Iter 1636: error - 43.96875000007837, weights: [

Iter 1891: error - 43.96875000000073, weights: [45.06249875  3.81250023]
Iter 1892: error - 43.968750000000725, weights: [45.06249876  3.81250022]
Iter 1893: error - 43.968750000000696, weights: [45.06249878  3.81250022]
Iter 1894: error - 43.968750000000696, weights: [45.06249879  3.81250022]
Iter 1895: error - 43.96875000000068, weights: [45.0624988   3.81250022]
Iter 1896: error - 43.96875000000067, weights: [45.06249881  3.81250022]
Iter 1897: error - 43.96875000000065, weights: [45.06249882  3.81250021]
Iter 1898: error - 43.96875000000065, weights: [45.06249883  3.81250021]
Iter 1899: error - 43.96875000000063, weights: [45.06249884  3.81250021]
Iter 1900: error - 43.968750000000625, weights: [45.06249885  3.81250021]
Iter 1901: error - 43.9687500000006, weights: [45.06249886  3.81250021]
Iter 1902: error - 43.9687500000006, weights: [45.06249887  3.8125002 ]
Iter 1903: error - 43.96875000000058, weights: [45.06249888  3.8125002 ]
Iter 1904: error - 43.96875000000058, weights: [4